rfcs/QueryLevelNullability.md

```diff
@@ -1,121 +1,187 @@
+    name: String
   }
-}
-```
-```swift
-struct GetBusinessNameQuery.Data.Business {
-  let name: String?
-}
-```
-In many cases, the client should error if the business `name` is null. While the schema can have the nullable
-fields for valid reasons (such as federation), the client wants to decide if it accepts a null value for]
-the result to simplify the client-side logic.  
+
+  # Document
+  query GetBusinessName($id: String!) {
+    business(id: $id) {
+      name
+    }
+  }
+  ```
+  At build time, Apollo generates the following Swift code (note: the code has been shortened for clarity).
+  ```swift
+  struct GetBusinessNameQuery {
+    let id: String
+
+    struct Data {
+      let business: Business?
+
+      struct Business {
+        /// Since the `Business.name` schema field is nullable, the corresponding codegen Swift property is `Optional`
+        let name: String?
+      }
+    }
+  }
+  ```
+  The query can then be fetched, and the resulting data handled, as follows:
+  ```swift
+  GraphQL.fetch(query: GetBusinessNameQuery(id: "foo"), completion: { result in
+    guard case let .success(gqlResult) = result, let business = gqlResult.data?.business else { return }
+
+    // Often, the client needs to provide a default value in case `name` is `null`.
+    print(business?.name ?? "null")
+  }
+  ```
+
+## 📜 Problem Statement
+
+The two modern languages used on mobile clients, Swift and Kotlin, are both non-null by default. By contrast, in a
+GraphQL type system, every field is nullable by default. From the [official GraphQL best practice](https://graphql.org/learn/best-practices/#nullability):
+
+> This is because there are many things that can go awry in a networked service backed by databases and other
+> services. A database could go down, an asynchronous action could fail, an exception could be thrown. Beyond
+> simply system failures, authorization can often be granular, where individual fields within a request can
+> have different authorization rules.
+
+This discrepancy means that developers using strongly-typed languages must perform `null` checks anywhere that GraphQL
+codegen types are used, significantly decreasing the benefits of codegen. In some cases, the codegen types may be so
+complex that a new model type which encapsulates the `null` handling is written manually.
+
+While the schema can have nullable fields for valid reasons (such as federation), in some cases the client wants to decide
+if it accepts a `null` value for the result to simplify the client-side logic. In addition, a syntax for this concept
+would allow codegen tooling to generate model types that are more ergonomic to work with, since the since the model
+type's properties would have the desired nullability
 
 ## 🧑‍💻 Proposed syntax
 
-It would make more sense if the client could express that `name` must be non-null _in the query itself_:
+The client can express that a schema field is required using the `!` syntax in the query definition:
```

Comment:
```suggestion
The client can express that a schema field is required by using the `!` syntax in the query definition:
```

Reply:
Could this syntax also be used on mutations? If so there a term that's used to describe both queries and mutations?

Reply:
I think "operations" is the term, but I can't think of how this syntax could be applicable to mutations. in any case we can keep it scoped to only "queries" for this proposal